---
title: "FinalProject"
author: "Alejandra"
date: "2024-05-11"
output: pdf_document
---

#load required packages
```{r message = FALSE}
library(dada2)
library(Biostrings)
library(ShortRead)
library(phyloseq)
library(dplyr)
library(BiMiCo)
library(ggplot2)
library(devtools)
library(MicEco)
library(vegan)
```
#load sequences
```{r}
path <- "Sequences"
list.files(path)
```
#read file names
```{r}
fnFs <- sort(list.files(path, pattern="_R1_001.fastq", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_R2_001.fastq", full.names = TRUE))
#extract file names
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
```

#inspect file quality of forward and reverse reads
```{r}
plotQualityProfile(fnFs[1:2])
plotQualityProfile(fnRs[1:2])
```
#filter and trim
```{r}
#place filtered files in filtered/ subdirectory 
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(220,220),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE) 
head(out)
```

#learn error rates of reads
```{r}
##learn error rates of forward and reverse reads
errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)
```

#visulaize error rate
```{r}
plotErrors(errF, nominalQ=TRUE)
plotErrors(errR, nominalQ=TRUE)
```

#will take reads and show how many sequences/species are in the sample
```{r}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
dadaRs <- dada(filtRs, err=errR, multithread=TRUE)
dadaFs[[1]]
```
#merge paired reads
```{r}
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)
# Inspect the merger data.frame from the first sample
head(mergers[[1]])
```

#construct sequence table to see how many sequences are present and length
```{r}
seqtab <- makeSequenceTable(mergers)
dim(seqtab)
# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))
```

#remove chimeras (two sperate reads that got smashed together)
```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
sum(seqtab.nochim)/sum(seqtab)
```

#track reads (which step lost reads)
```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
```

#save setab.nochim as an R file
```{r}
save(seqtab.nochim, file= "RData/seqtab.nochim.RData")
```

#load seqtab.nochim
```{r}
load("RData/seqtab.nochim.RData")
```

#asign taxonomy
```{r}
taxa <- assignTaxonomy(seqtab.nochim, "silva_nr99_v138.1_wSpecies_train_set.fa.gz", multithread=TRUE)
```

```{r}
save(taxa, file = "RData/taxa.RData")
```

#load taxa and seqtab.nochim
```{r}
load("RData/taxa.RData")
load("RData/seqtab.nochim.RData")
```

#import metadata
```{r}
metadata <- read.csv("metadatanocontrol.csv", header=TRUE, row.names = 1)
```

#create physeq object
```{r}
physeq <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows = FALSE), 
                 sample_data(metadata), 
                 tax_table(taxa))
physeq
```

```{r}
#convert from raw to abundance so its easier to comapre
physeq <- transform_sample_counts(physeq, function(abund) 1*(abund>0))

#visualize to data
physeq
```
#remove the sequence itselt and replace with ASV
```{r}
##this allows it to be easier to read, replaces the raw data
dna <- Biostrings::DNAStringSet(taxa_names(physeq))
names(dna) <- taxa_names(physeq)
physeq <- merge_phyloseq(physeq, dna)
taxa_names(physeq) <- paste0("ASV", seq(ntaxa(physeq)))
physeq
```
#remove mitochondria and phloroplast mathces, remove all non bacterial sequences
```{r}
#stictly use bacteria 16S rRNA, 
physeq <- physeq %>% subset_taxa( Family!= "Mitochondria" | is.na(Family) & Order!="Chloroplast" | is.na(Order))
physeq
```

#remove all non bacterial sequences
```{r}
physeq<-rm_nonbac(physeq)
physeq
```

#save physeq objects and load 
```{r}
save(physeq, file= "RData/physeq.RData")
```

```{r}
load("RData/physeq.RData")

```

#plot bar grpah based on phylum
```{r}
plot_bar(physeq, fill = "Phylum") + geom_bar(aes(color=Phylum, fill=Phylum), stat="identity", position="fill")
```

#create a barplot of relative abundance
```{r}
#convert to relative abundance
physeq_relabund <- transform_sample_counts(physeq, function(x) x / sum(x))

#barplot
plot_bar(physeq_relabund, fill = "Phylum") + geom_bar(aes(color=Phylum, fill=Phylum), stat="identity", position="fill") + facet_wrap(~site, scales= "free")
##can change based on the column name in metadata in facet_wrap(~columnName)
```

#plot alpha diversity based on bird
```{r}
plot_richness(physeq, x="bird", color= "speciessite", measures=c("Observed", "Simpson", "Shannon"))
##Simpson(less sensitive, will be more custered together) and Shannon(more sensitive to rare taxa) take into consideration relative abundance each species.
```

#plot alpha diversity based on site
```{r}
plot_richness(physeq, x="species", color= "site", measures=c("Observed", "Simpson", "Shannon"))
##Simpson(less sensitive, will be more custered together) and Shannon(more sensitive to rare taxa) take into consideration relative abundance each species.
```
#plot alpha diversity based on sex
```{r}
plot_richness(physeq, x="bird", color= "species", measures=c("Observed", "Simpson", "Shannon"))
##Simpson(less sensitive, will be more custered together) and Shannon(more sensitive to rare taxa) take into consideration relative abundance each species.
```

#creating file with statistics of alpha diversity 
```{r}
alphadiv <- estimate_richness(physeq, measures=c("Observed", "Simpson", "Shannon"))
write.csv(alphadiv, "alpha.csv")
```

#load file that was manually made that has metadata and alpha diversity statistics
```{r}
alphameta <- read.csv("alphameta.csv")
```

#perform normaility test on diversity index using qq plot and shapiro
```{r}
#observed stats
qqnorm(alphameta$Observed)
qqline(alphameta$Observed)
#shannon index stats
qqnorm(alphameta$Shannon)
qqline(alphameta$Shannon)
#simpson stats
qqnorm(alphameta$Simpson)
qqline(alphameta$Simpson)


#shaprio test
observed <- shapiro.test(alphameta$Observed)
shannon <- shapiro.test(alphameta$Shannon)
print(observed)
print(shannon)
```

#performe statics on site
```{r}
#perform wilcox tests for each biodiversity index
test_observed<- wilcox.test(alphameta$Observed ~ alphameta$site)
test_shannon <- wilcox.test(alphameta$Shannon ~ alphameta$site)
test_simpson <- wilcox.test(alphameta$Simpson ~ alphameta$site)
print(test_observed)
print(test_shannon)
print(test_simpson)
```
#performe statics on species
```{r}
#perform wilcox tests for each biodiversity index
test_observed<- wilcox.test(alphameta$Observed ~ alphameta$species)
test_shannon <- wilcox.test(alphameta$Shannon ~ alphameta$species)
print(test_observed)
print(test_shannon)
```
#performe statics on speciessite
```{r}
#perform kruskal tests for each biodiversity index
test_observed<- kruskal.test(alphameta$Observed ~ alphameta$speciessite)
test_shannon <- kruskal.test(alphameta$Shannon ~ alphameta$speciessite)
print(test_observed)
print(test_shannon)
```
#doing a kruskal-wallis on birds and diversity indices
```{r}
test_observed <- kruskal.test(alphameta$Observed ~ alphameta$bird)
test_shannon <- kruskal.test(alphameta$Shannon ~ alphameta$bird)
print(test_observed)
print(test_shannon)
```


#remove taxa with relative abundance <0.05%
```{r}
minTotRelAbun = .00005
x = taxa_sums(physeq)
keepTaxa = (x / sum(x)) > minTotRelAbun
physeqprune = prune_taxa(keepTaxa, physeq)
physeqprune
```


#bray curtis caculation, 0; exactly the same, 1; very diverse 
```{r message=FALSE}
set.seed(666)
dist = phyloseq::distance(physeqprune, method="bray", weighted=TRUE)
ordination = ordinate(physeqprune, method="NMDS", distance=dist)
```
#bray curtis site plot
```{r}
braysite=plot_ordination(physeqprune, ordination, color="site") + 
  theme_classic() +
  theme(strip.background = element_blank()) + stat_ellipse(aes(group=site))
braysite
```

#bray curtis species plot
```{r}
brayspecies=plot_ordination(physeqprune, ordination, color="species") + 
  theme_classic() +
  theme(strip.background = element_blank()) + stat_ellipse(aes(group=species))
brayspecies
```
#bray curtis speicessite plot
```{r}
brayspeciessite=plot_ordination(physeqprune, ordination, color="speciessite") + 
  theme_classic() +
  theme(strip.background = element_blank()) + stat_ellipse(aes(group=speciessite))
brayspeciessite
```

#bray curits statistics
```{r}
#on species
adonis2(dist ~ sample_data(physeqprune)$species)
#on site
adonis2(dist ~ sample_data(physeqprune)$site)
```

#bray curtis species and site stats
```{r}
adonis2(dist ~ sample_data(physeqprune)$speciessite)
ps.disper<-betadisper(dist, sample_data(physeqprune)$speciessite)
permutest(ps.disper, pair=TRUE)
```