---
title: "Microbiome dada2"
author: "Alejandra"
date: "2024-04-16"
output: pdf_document
---

#load required (dada2) packages
```{r}
library(dada2)
```

#load sequences
```{r}
path <- "sequences"
list.files(path)
```

#read file names
```{r}
# Forward fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq
fnFs <- sort(list.files(path, pattern="_R1_001.fastq", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
```

#inspect read quality
```{r}
plotQualityProfile(fnFs[1:2])
```

#filter and trim
```{r}
# Place filtered files in filtered/ subdirectory
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
names(filtFs) <- sample.names

#quality decreses at 120
out <- filterAndTrim(fnFs, filtFs,  truncLen=c(120),
              maxN=0, maxEE=c(2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE
head(out)
```

#learn error rates
```{r}
errF <- learnErrors(filtFs, multithread=TRUE)

#visulaize error rate
plotErrors(errF, nominalQ=TRUE)
```

#create sample inference (identifying number of unique sequences)
```{r}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
```

#constrict sequence table
```{r}
#change mergers to dadaFs because no paired reads
seqtab <- makeSequenceTable(dadaFs)
dim(seqtab)

# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))
```

#remove chimeras
```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
```

#track reads
```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out,  sapply(dadaFs, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. remove reverse reads and merge
colnames(track) <- c("input", "filtered", "denoisedF", "nonchim")
rownames(track) <- sample.names
head(track)
```
#save setab.nochim as an R file
```{r}
save(seqtab.nochim, file= "RDatata/seqtab.nochim.RData")
```

#load seqtab.nochi, to start here
```{r}
load("RData/seqtab.nochim.RData")
```

#assign taxanomy
```{r}
#downlaod the file https://zenodo.org/records/4587955

